### ISSUE TITLE
fonts: looking for some mac users (windows testers also welcome)

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [arkenfox/user.js#1211]
Hey guys and gals

Here's some background, more to follow next comment

on TZP I am trying to improve perf. A large chunk of time is taken up by the font detection test. Aside from a little time taken to setup the font element, and some differences for a found font vs one that isn't there (it takes a millisecond or so more to load a font that to not load it: so the less detectable fonts the slightly faster), it is essentially a case of limiting the fonts tested for for near-linear gain.

And not all OSes have the same lists or numbers. I have a master list for each OS, and I know the OS via JS

On windows for example on my machine (cold browser starts, zero cache etc)
- the full font list of `485`, it used to take ~400+/1200ms .. call it ~30%
- the reduced list of `249`, it takes ~200/1000ms (~20%)
- the RFP base list of `117`, it takes ~120/800ms (~15%)  (tested with RFP off)
- on TB with `81` its down to 90ms (timed with RFP off)

---

### Ways to reduce the number tested for

e.g. windows master list was 485 fonts

split off TB bundled fonts (I can 100% reliably detect TB)
- this allows me to remove the fonts bundled by tor browser: currently at 5 (every bit helps)
- 👍 now down to `480`

discard font styles
- e.g. we can test for the font family `AngsanaUPC` but do not need to test for styles `AngsanaUPC Bold`,`AngsanaUPC Bold Italic`,`AngsanaUPC Italic`, as the styles will never be returned as found even if you have them
- it's not always clear what is a style and what is a family e.g. `Arial` and `Arial Black`, `Arial Narrow`
- at least on windows it seems as if `x Bold`, `x Bold Italic` and `x Italic` are family styles if the font family name `x` exists, and `narrow` and `black` and `light` and `condensed` etc are families
- there seem to be exceptions to this rule of thumb, and ultimately it takes testing to be 100% sure
- 👍 now down to `249`

only test base fonts if RFP is on (and you're not Tor Browser, and FF is version 80 or higher)
- unfortunately I can only do this for windows and mac
- linux is only for some distros, unless I can always detect the distro, which may be too hard to do given desktop enviroments
- 👍 now down to `117`

Tor Browser
- use a list based on TB's whitelist pref: windows and mac only
- 👍 this brings TB down to `81` (including the bundled fonts)

possibly
- don't check for fonts that we `expect` should be there: e.g. all windows will have Arial, and Times New Roman etc. Some system fonts can't be deleted (a small handful), but a lot can, and I'm not sure if I really want to compromise the test by making assumptions - but something to think about
- detect win7 vs win10 (not sure about win8/8.1) and reduce the lists even more. Not sure about this, as users can add fonts, e.g from bundled apps, manually

windows: so currently on TZP ... FF users = `249` fonts to check, RFP users (FF80+) = `117`, and TB is `81` fonts 🎉 

will continue next post

**Comment by Thorin-Oakenpants** [comment]
reserved

**Comment by TearJoy** [comment]
<details>
  <summary>min + RFP/TB</summary>
  
![image](https://user-images.githubusercontent.com/81557214/125404133-d229dd00-e3a5-11eb-896a-24644ff48756.png)

</details>

<details>
  <summary>styles</summary>
  
![image](https://user-images.githubusercontent.com/81557214/125404275-fa194080-e3a5-11eb-9836-0f1bd332b4e3.png)


</details>


OS Version: Windows 11 Insider Dev Build 22000.65

**Comment by Thorin-Oakenpants** [comment]
@TearJoy 👍 Thanks:) That's interesting, because I also got a similar (IDK if it was exactly the same) result several days ago in my win10 VM, this was before I output the found fonts. I did some quick manual diff spotting - and I have a note to retest and follow up on it - but I was very tired at the time

See the `101` result, where you have RFP enabled - see `Franklin Gothic Medium` is missing and yet it is in the test's `base` list, also shown [here](https://searchfox.org/mozilla-central/source/gfx/thebes/StandardFonts-win10.inc) in kBaseFont. I know you have it, because it shows up in your `103` style result, where RFP is off.

So I wasn't mistaken it in my tired state, by the looks of it. I think this might be a bug.

**Comment by TearJoy** [comment]
No problem! Always happy to contribute to this project whenever I can.

**Comment by Onfroygmx** [comment]
Hello,

Results for MacOs Big Sur 11.4:

min + RFP/TB
![min + RFP](https://user-images.githubusercontent.com/25054585/125426692-cea3c469-24f2-4ebe-80a2-94e5feec71a6.png)

Styles:
![2021-07-13 at 11-15-47 styles](https://user-images.githubusercontent.com/25054585/125426844-81fdea45-fb61-435d-9fa8-cc2d9f762564.png)

Note that the results may vary depending on which version of MacOs you are on, they like to add new fonts.

sd

**Comment by stan1408** [comment]
Results for Windows 7
![min](https://user-images.githubusercontent.com/32367438/125477888-647ffc8e-07a8-434e-a4f9-b3f519607a41.jpg)
![styles](https://user-images.githubusercontent.com/32367438/125477946-1eff47da-0b49-4ac4-877c-ddcc0f7e210b.jpg)

**Comment by moratismo** [comment]
Results for Windows 7 SP1
 
<details>
<summary>min + RFP/TB</summary>

![image](https://user-images.githubusercontent.com/80642566/125494906-de36932b-0e58-4fbd-825b-839a61cb0ee0.png)
</details>

<details>
<summary>styles</summary>

![image](https://user-images.githubusercontent.com/80642566/125498419-bd00d521-2b54-4f6c-ad48-e5221a52b856.png)
</details>

**Comment by GoodChickenParmesan** [comment]
Hi nice to meet you guys. Just made an account.
Mac-OS Big Sur 11.4

Min + RFP/TB
![Minutes + RFP:TB](https://user-images.githubusercontent.com/87397152/125539257-02aa07a7-a2a9-4f81-acdc-b3c733fd9fdf.png)

Styles
![Styles](https://user-images.githubusercontent.com/87397152/125539299-3b9ac617-68e9-4f31-abe6-f76ec40629a4.png)

**Comment by Thorin-Oakenpants** [comment]
Thanks so much you guys. Good to see the styles tests so far on Mac show I haven't removed too many items

Interesting @Onfroygmx has a blue checkbox and @GoodChickenParmesan has a purple one :) Not fingerprintable, AFAIK widgets are somewhat protected more than other elements, but I should check

@GoodChickenParmesan welcome .. 🍰 🍺

**Comment by ggl-wolfy** [comment]
my results on macOS Big Sur 11.4

min + RFP/TB
<img width="462" alt="Screen Shot 2021-07-14 at 1 06 08 pm" src="https://user-images.githubusercontent.com/58021228/125554923-7ec6ca31-901f-4ac2-b3fb-d77384fa5e6c.png">

not sure why my results don't agree with the others?

styles
<img width="385" alt="Screen Shot 2021-07-14 at 1 08 30 pm" src="https://user-images.githubusercontent.com/58021228/125555003-700f7127-7e51-42e5-802a-6e4af8771d74.png">

**Comment by Thorin-Oakenpants** [comment]
> not sure why my results don't agree with the others

it does: your RFP results are the same: `190` found from the `min` list, hash `2321f206...`

What's weird is that the min test (which checks you have RFP on) also finds `Osaka-Mono`, which is not in the min list, and shouldn't be - [see this list](https://searchfox.org/mozilla-central/source/gfx/thebes/StandardFonts-macos.inc)

**Comment by ggl-wolfy** [comment]
Seems like I do have Osaka-Mono installed, although I have no recollection of installing it.

<img width="999" alt="Screen Shot 2021-07-14 at 5 46 52 pm" src="https://user-images.githubusercontent.com/58021228/125583894-e8a48ca6-b126-4c3d-8229-063b98be5c42.png">

**Comment by Thorin-Oakenpants** [comment]
Thanks :) That it was detected in RFP mode but it's not on the kBaseFonts list is perplexing ... and the only reason I can think of, is that "bundled" fonts are auto-included - otherwise it's a bug somewhere (either how the detection code works, which seems unlikely given you have the font) or upstream at Mozilla

a bundled font is one that resides in the Mozilla font directory - may someone else can tell you exactly where that is if you don't know (I don't, sorry) - it would be nice to check that this is or isn't the cause

side note: I'm not a font expert by all means, but I wonder if the list should be `Osaka Regular` and not `Osaka`. I wish I had a mac to test on *sigh*

**Comment by Onfroygmx** [comment]
If you tell me approximatively what you want I can test it with a clean profile (with or without user.js), I'll check the comments every hour today.

Onf

**Comment by Thorin-Oakenpants** [comment]
@Onfroygmx ... at a bare minimum us all where Firefox has it's font directory. Other than that, nothing to test really - I just wish I had a mac, so I could see perf, and other things, like how retina + devicePixelRatio changes things etc.

If you really wanted to, you could install Osaka and repeat the test - and then uninstall it ... but I don't want to put anyone to any trouble, and I think I have enough to go on, especially if we can just confirm ggl-wolf doesn't have it in his font dir

If I had my own mac
- I would install every single optional language pack
- I would make sure I had every single font on the lists
- I would expand the list to include all sorts of variations like `Osaka-Regular` etc and I could confirm I had the fonts and their names because I would have the OS in front of me
- Then I would do a full test and anything not found must be a family style we can ignore
- I could also use the whitelist pref with just a single made up font in it, to see what fonts are always reported regardless: fonts are required in an app, you see - the UI and needs to function and some system fonts at least on windows are always returned
- And I would install Tor Browser and test that as well

But I don't expect anyone to do that ...

**Comment by Onfroygmx** [comment]
First remarks:

- There does not seem to be a specific font folder for FF, it picks up the system fonts: 
  "/Users/$USER/Library/Fonts"
- After install of an OSAKA font it does not seem to be picked-up directly by FF (Will retest after a restart later on).
- Modifications since yesterday tests and upgrade to FF 90:

<details>
  <summary>min + RFP/TB</summary>
![RFP:TB_new](https://user-images.githubusercontent.com/25054585/125598518-aa85b221-2e09-4e98-ac2d-f9d71aee1fc9.png)

</details>

<details>
  <summary>styles</summary>
![styles_new](https://user-images.githubusercontent.com/25054585/125598731-5f654efd-6bed-490d-a9b2-e2a92b652746.png)

</details>

Some fonts have disappeared from the min results.

Onf

**Comment by Thorin-Oakenpants** [comment]
> There does not seem to be a specific font folder for FF

Hmm OK. I know Tor Browser bundle fonts (IDK about mac, but on windows it's basically a portable app), FF portable has one, and windows and linux have the capability .. can't see why Mac doesn't. e.g.
- windows : `...\Program Files\Mozilla Firefox\fonts`
- windows portable: `...\FirefoxPortable\App\firefox64\fonts`
- windows TB: `...\Tor Browser\Browser\fonts`

> Some fonts have disappeared from the min results.

I removed two fonts, as Mozilla don't bundle an emoji font for macs - that's why always is zero. Your result though is the same as before .. 190 found, same hash

**Comment by Thorin-Oakenpants** [comment]
TIL ... they're called ["faces"](https://www.w3.org/TR/css-fonts-4/#font-families), not "styles"

**Comment by fxbrit** [comment]
tests with stock fonts:
<details>
 <summary>min</summary>

![min](https://user-images.githubusercontent.com/34379516/125602522-4de7df6e-248a-49f1-8d8c-e3ae961377b1.png)

</details>
<details>
 <summary>full</summary>

![full](https://user-images.githubusercontent.com/34379516/125602212-da86eee5-08ff-4ffb-a668-c2995a49c165.png)

</details>

then I installed all available fonts, including chinese ones, osaka and all the possible variations (faces) of it.
<details>
 <summary>proof</summary>

<img width="268" alt="fonts" src="https://user-images.githubusercontent.com/34379516/125603242-bcadc439-dd1a-45ec-a650-7e1334b9eba8.png">

</details>
<details>
 <summary>min</summary>

![min-all fonts](https://user-images.githubusercontent.com/34379516/125603619-18d917fe-b29f-48e6-9d1c-c3c093d8cfeb.png)

</details>
<details>
 <summary>full</summary>

![full-all fonts](https://user-images.githubusercontent.com/34379516/125603658-75a65553-f99e-445d-aafc-f7cbee596803.png)

</details>

how the test was run:
- macOS 11.4
- vanilla firefox 90.0, fresh install
- rfp on for 'min' and off for 'full', as required

**Comment by ggl-wolfy** [comment]
I'm not sure if this is helpful, but after I disabled the Osaka font through MacOS system "Font Book.app", the max hash is now the same as the min hash with 190 fonts found.

**Comment by Thorin-Oakenpants** [comment]
OK, thanks guys ever so much. So that's two mac users with `Osaka*` and the font visibility pref (or RFP which uses it), lets the font still get detected when it shouldn't. So that's a bug: either me or upstream ... got all I need for that issue now

@ggl-wolfy : yup, that sounds right. The issue isn't you having the font installed or enabled or whatnot, it is that the above sentence I just typed out - some sort of bug

more freebies, help yourselves -> 🥧 🍸 🥮 🍻 🍕 

Hmm ... full 260, and the medium is 483 tested .. I wonder if there is more I can trim out .. another day, I'm beat

**Comment by Onfroygmx** [comment]
Don't forget it could also just be a Big Sur bug, they changed a lot with all the basic settings (system is installed on a secured snapshot now). You would need to have someone with Catalina or below to pinpoint the origin of the bug for sure.

**Comment by Thorin-Oakenpants** [comment]
@fxbrit are you on Big Sur?

**Comment by Onfroygmx** [comment]
To give some insight on what I said:
https://eclecticlight.co/2021/07/15/is-it-safe-yet-to-upgrade-to-big-sur/

> The major architectural change from Catalina is that Big Sur’s System volume isn’t just mounted read-only, it’s not mounted at all! ... Your Mac boots not from the System volume, but from that sealed snapshot. ... This has pervasive effects. ... So what Apple puts on the System volume, in the way of all its standard fonts, for example, is what you’ve got.

Onf.

**Comment by fxbrit** [comment]
some more tests. again macOS 11.4 and fresh install of firefox 90.0.

<details>
 <summary>min, default list of fonts</summary>

![min-default-list](https://user-images.githubusercontent.com/34379516/126035303-5e151117-adaa-443a-a7dd-c84afe2c563e.png)

</details>

<details>
 <summary>full, default list of fonts</summary>

![full-default-list](https://user-images.githubusercontent.com/34379516/126035306-7018dcd7-319e-408b-994b-da5fce8fa676.png)

</details>

<details>
 <summary>min, after downloading all fonts in font book</summary>

![min-all](https://user-images.githubusercontent.com/34379516/126035316-f722f0b4-938f-4a9b-99d8-4f04220e8202.png)

</details>

<details>
 <summary>full, after downloading all fonts in font book</summary>

![full-all](https://user-images.githubusercontent.com/34379516/126035321-8c7d81b2-830c-495b-89f3-2d808b5e344f.png)

</details>

**Comment by Thorin-Oakenpants** [comment]
^ Tor Browser mac whitelisting is set up

**Comment by Onfroygmx** [comment]
Mac OS 11.4, Tor 10.5.2:

<details>
<summary>min + RFP/TB</summary>
![tormin](https://user-images.githubusercontent.com/25054585/126037859-1ae22c4c-d314-42a7-b766-a884ba4b28ec.png)
</details>

<details>
<summary>full</summary>
![torfull](https://user-images.githubusercontent.com/25054585/126037864-12cfb090-501a-4848-b79a-43c28ef3d17e.png)
</details>

**Comment by abrahamjuliot** [comment]
Mac OS 10.15 (VM)

<details>
<summary>min</summary>

![image](https://user-images.githubusercontent.com/6946045/126086957-6942b4fe-35b2-4028-9106-ee23940398f9.png)
</details>
<details>
<summary>full</summary>

![image](https://user-images.githubusercontent.com/6946045/126087158-9fdb6737-b587-4c9c-bf19-77ed1259bb4c.png)

![image](https://user-images.githubusercontent.com/6946045/126087051-2899dce0-75d5-47cc-b780-62ed9ba89477.png)
</details>

**Comment by Thorin-Oakenpants** [comment]
interesting missing `Symbol` compared to other 10.15's

**Comment by rampaa** [comment]
<details>
  <summary>Min</summary>

![image](https://user-images.githubusercontent.com/25622653/126099290-a1794841-d342-4443-84de-be130d9796bb.png)
</details>

<details>
  <summary>Full</summary>

![image](https://user-images.githubusercontent.com/25622653/126099242-4fa0cf7a-8206-4aac-a85e-5259ca846e44.png)
</details>

**Comment by Thorin-Oakenpants** [comment]
thanks @rampaa guess those two `mediums` were the only ones in face - will fix 💋

**Comment by GoodChickenParmesan** [comment]
MacOs Big Sur 11.4
Tor Browser 10.5.2

![min + RFP:TB on Tor](https://user-images.githubusercontent.com/87397152/126167792-c94e1a80-539e-4326-8396-a84892b70b4c.png)

![full on Tor](https://user-images.githubusercontent.com/87397152/126167808-1ba8fed0-2b64-47e6-8d6c-06f066c0032e.png)

**Comment by Thorin-Oakenpants** [comment]
Thanks. FYI, Tor Browser tests for full are a bit useless, easier to just run full tests in Firefox: the full test is to check that anything in min is not blocked when protection is on .. (it reports this when you have run both tests) .. and to see what "font names" I can discard as part of a "parent" font famly

**Comment by Thorin-Oakenpants** [comment]
thanks everyone. I'm going to close this, although I plan to change up the mac fonts

I need to add a lot of 10.15 fonts not listed (currently I had only added new 10.15 base fonts). This makes a very big "medium" list (the one used by TZP), so I'm going to create an `fntExtra` array and move items into it (e.g. legacy since 10.12 i.e. allowed in documents, and some that can moved as expected or likely). TZP just needs to demonstrate that the font protection is working, not necessarily detect *every* font. That's what this font list test page can do instead, check there are no leaks or bugs (it already found two) - FYI: the Osaka-Mono font was a bug which has been fixed in FF91+
